---
title: "THRA Isoform 1 and Isoform 2 Relative Expression"
output:
  html_document:
    toc: True
    toc_float: True
    df_print: paged
---

INTRODUCTION to the Experiment

We applied three complementary RNA-seq methods on a heterogeneous collection of 300 human samples, including 45 tissues, 162 cell types, and 93 cell lines (cancer). From these samples, we generated strand-specific small RNA (298 samples), polyA (295 samples) and total RNA (296 samples) libraries that were sequenced at a median depth of 13 M, 60 M paired-end, and 125 M paired-end reads respectively, resulting in a total of 125 Billion reads.
First-Stranded paired-end reads. 75bp x 2. Read counts of THRA exon 9a (Chr17:40089333) and exon 9b (Chr17:40089334)

```{r}
#load libraries
library(tidyverse)
library(readr)
library(ggplot2)
library(plotly)
library(matrixStats)
library(ggrepel)
library(scales)
library(readxl)
library(dplyr)

```


```{r}
## set paths for output figure
path_plots <- "~/OneDrive - Charité - Universitätsmedizin Berlin/Schuelke_Lab/EG15_RNA_Seq/GEO_GSE138734/THRA_isoform_Expression_Coverage"

## load metadata file

metadata <- read.delim("GEO_GSE138734_metadata.txt", sep=",")
metadata_filtered <- select(metadata, Run, "Sample" =source_name, sample_type, organ_system) 


##Load bedtools outputs
file_links_bedtools <- list.files(path= "~/OneDrive - Charité - Universitätsmedizin Berlin/Schuelke_Lab/EG15_RNA_Seq/GEO_GSE138734/bedtools_outputs" , pattern = "*.txt", full.names=T)


# initialize an empty dataframe
data_bedtools <- data_frame("V1"=character(),
                   "V2"=integer(),
                   "V3"=integer(),
                   "V4"=character(),
                   "V5"=double(),
                   "Run"=character()) 

for (x in file_links_bedtools) {
  table_sample <- read.delim(x, header = FALSE) ## read table
  basename_sample <- basename(x)
  table_sample <- mutate(table_sample, "Run"=basename_sample)
  treshold_value <- filter(table_sample, V4 != "THRA") %>%
    select(V5)
  
  # if statement to filter out samples that have 0 coverage in both exon 9a and 9b
  if (sum(treshold_value)  > 20) {
    data_bedtools <- bind_rows(data_bedtools, table_sample)
  }
    
  
}

#extract the SSR number and merge with the Sample name
data_bedtools <- data_bedtools %>%
  separate("Run","Run", "_") %>%
  left_join(metadata_filtered, by="Run")

data_bedtools

##Load mosdepth outputs

file_links_mosdepth <- list.files(path= "~/OneDrive - Charité - Universitätsmedizin Berlin/Schuelke_Lab/EG15_RNA_Seq/GEO_GSE138734/mosdepth_outputs" , pattern = "*.bed", full.names=T)


# initialize an empty dataframe
data_mosdepth <- data_frame("V1"=character(),
                   "V2"=integer(),
                   "V3"=integer(),
                   "V4"=character(),
                   "V5"=double(),
                   "Run"=character()) 

for (x in file_links_mosdepth) {
  table_sample <- read.delim(x, header = FALSE) ## read table
  basename_sample <- basename(x)
  table_sample <- mutate(table_sample, "Run"=basename_sample)
  treshold_value <- filter(table_sample, V4 != "THRA") %>%
    select(V5)
 # treshold_value <- treshold_value$V5
  
  # if statement to filter out samples that have 0 coverage in both exon 9a and 9b
  if (sum(treshold_value)  > 20) {  ## generally should be at least 4
    data_mosdepth <- bind_rows(data_mosdepth, table_sample)
  }
    
  
}

#extract the SSR number and merge with the Sample name
data_mosdepth <- data_mosdepth %>%
  separate("Run","Run", "_")%>%
  left_join(metadata_filtered, by="Run")

data_mosdepth

```

```{r}
## Let's rearrange the data in a useful way
## I will keep track of Run as the replicate identifier
data_mosdepth <- data_mosdepth %>%
  select(Sample, Run, sample_type, Isoform = V4, Reads_count =V5) %>%
  spread(key="Isoform", value="Reads_count") %>%
  mutate("Package" = as.factor("mosdepth")) 

data_mosdepth

data_bedtools <- data_bedtools %>%
  select(Sample,Run,sample_type, Isoform = V4, Reads_count =V5) %>%
  distinct() %>%
  spread(key="Isoform", value="Reads_count")%>%
  mutate("Package" = as.factor("bedtools")) %>%
  filter(Run %in% data_mosdepth$Run) ## makes sure we take only samples that were filtered already

data_bedtools

```

```{r}
## Let's merge the two dataframes into one

dataset <- full_join(data_mosdepth, data_bedtools) 

dataset
```



```{r}
## Let's calkculate THRA1 (counts of 9b) and THRA2 (9a-9b)

dataset <- dataset %>%
  mutate("THRA1"= dataset$`9b`) %>%
  mutate("THRA2"=dataset$`9a`-dataset$`9b`) %>%
  rename("Read_counts_9a" = "9a") %>%
  rename("Read_counts_9b" = "9b") %>%
  relocate(Package, .after=Sample)

dataset

```
```{r}
## Let's add the final calculations

dataset_final <- dataset %>%
  mutate("delta_A1vsA2" = THRA1 - THRA2) %>%
  mutate("THRA1_Percentage" = round(THRA1/Read_counts_9a*100)) %>%
  mutate("THRA2_Percentage" = round(100-THRA1_Percentage)) %>%
  mutate("THRA1_higher" = THRA1 > THRA2) %>%
  mutate("delta_percentage" = THRA1_Percentage - THRA2_Percentage)

dataset_final
  
```






```{r}
### Lets compare the read counts of exon 9a between the 2 tools
graph_data_exon9a <- dataset_final %>%
  filter(sample_type == "tissue") %>%
  select(Sample, Package, Read_counts_9a)


ggplot(graph_data_exon9a, aes(x = Sample, y = Read_counts_9a, fill = Package)) +
  geom_boxplot( width=0.5, position=position_dodge(width=0.7)) +
  scale_y_continuous("Read counts") +
  scale_fill_manual("", values = c("bedtools" = "darksalmon", "mosdepth" = "#56B4E9")) +
  ggtitle("Tools comparison - Exon 9a ") +
  theme_light(base_size = 12) +
  scale_x_discrete(guide = guide_axis(angle = -35)) +
  theme(axis.title.x=element_blank())

#ggplotly()

## save plot
ggsave("Read_counts_9a_tool_comparison.png" , device=png, dpi = 600, path = path_plots, bg = "transparent", width = 20, height = 15, units = "cm" )
```




```{r}
### Lets compare the read counts of exon 9b between the 2 tools
graph_data_exon9b <- dataset_final %>%
  select(Sample, Package, Read_counts_9b)


ggplot(graph_data_exon9b, aes(x = Sample, y = Read_counts_9b, fill = Package)) +
  geom_boxplot( width=0.5, position=position_dodge(width=0.7)) +
  scale_y_continuous("Read counts") +
  scale_fill_manual("", values = c("bedtools" = "darksalmon", "mosdepth" = "#56B4E9")) +
  ggtitle("Tools comparison - Exon 9b ") +
  theme_light(base_size = 12) +
  scale_x_discrete(guide = guide_axis(angle = -35)) +
  theme(axis.title.x=element_blank())

#ggplotly()
## save plot
ggsave("Read_counts_9b_tool_comparison.png" , device=png, dpi = 600, path = path_plots, bg = "transparent", width = 20, height = 15, units = "cm" )
```




```{r}
### heatmap with continuos delta THRA1vsA2 using mosdepth
heatmap_data_mosdepth <- filter(dataset_final, Package == "mosdepth") %>%
  group_by(Sample, Package) %>%
  summarize(delta_A1vsA2=round(mean(delta_A1vsA2)))

y_max_mosdepth <- max(heatmap_data_mosdepth$delta_A1vsA2)
y_min_mosdepth <- min(heatmap_data_mosdepth$delta_A1vsA2)

ggplot(heatmap_data_mosdepth, aes(1, reorder(Sample,delta_A1vsA2), fill=delta_A1vsA2)) +
  geom_tile()+
  ggtitle("THRA isoform expression pattern (THRA1/THRA2) - mosdepth") +
  theme_light(base_size = 12)+
  theme(axis.title.y=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_fill_gradient2('delta_reads', limits=c(y_min_mosdepth, y_max_mosdepth), breaks = c( y_min_mosdepth, 0, y_max_mosdepth),  low = "#56B4E9", high = "darksalmon", guide="colorbar")

ggsave("Heatmap_THRA1vsA2_mosdepth.png" , device=png, dpi = 600, path = path_plots, bg = "transparent", width = 20, height = 15, units = "cm" )
```
```{r}
### heatmap with AUT AUT delta THRA1vsA2 using mosdepth
heatmap_data_mosdepth <- heatmap_data_mosdepth %>%
  mutate("THRA1_higher" = delta_A1vsA2 > 0)

ggplot(heatmap_data_mosdepth, aes(1, reorder(Sample,delta_A1vsA2), fill=THRA1_higher)) +
  geom_tile()+
  ggtitle("THRA isoform expression pattern (THRA1/THRA2) - mosdepth") +
  theme_light(base_size = 12)+
  theme(axis.title.y=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_fill_manual(values = c("TRUE"= "darksalmon", "FALSE" = "#56B4E9"))

ggsave("Heatmap_THRA1vsA2_AUT_AUT_mosdepth.png" , device=png, dpi = 600, path = path_plots, bg = "transparent", width = 20, height = 15, units = "cm" )
```

```{r}
### heatmap of percentages using mosdepth
heatmap_data_mosdepth <- filter(dataset_final, Package == "mosdepth", sample_type == "tissue") %>%
  group_by(Sample, Package) %>%
  summarize(THRA1_Percentage=round(mean(THRA1_Percentage)))

ggplot(heatmap_data_mosdepth, aes(1, reorder(Sample,THRA1_Percentage), fill=THRA1_Percentage)) +
  geom_tile()+
  ggtitle("THRA isoform expression pattern (THRA1/THRA2) - mosdepth") +
  theme_light(base_size = 12)+
  theme(axis.title.y=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_fill_gradient2('% of THRA1', limits=c(0, 100), breaks = c( 0, 50, 100),  low = "#56B4E9", high = "darksalmon", guide="colorbar",  midpoint = 50)

  ggsave("Heatmap_THRA1vsA2_mosdepth_percentages.png" , device=png, dpi = 600, path = path_plots, bg = "transparent", width = 20, height = 15, units = "cm" )
```

```{r}
### Lets plot the percentage in a different way, so that the actual percentage is visible

heatmap_data_percentage <- filter(dataset_final, Package == "mosdepth" , sample_type == "tissue")%>%
  group_by(Sample) %>%
  summarize(THRA1_Percentage=round(mean(THRA1_Percentage))) %>%
  mutate("THRA1_higher" = THRA1_Percentage >50 ) %>%
  mutate("delta_percentage" = THRA1_Percentage - (100 - THRA1_Percentage))

heatmap_data_percentage <- heatmap_data_percentage[order(heatmap_data_percentage$THRA1_Percentage), ]  # sort
heatmap_data_percentage$Sample <- factor(heatmap_data_percentage$Sample, levels = heatmap_data_percentage$Sample)  # convert to factor to retain sorted order in plot.


# Diverging Barcharts
ggplot(heatmap_data_percentage, aes(x=Sample, y=THRA1_Percentage, label=THRA1_Percentage)) + 
  geom_bar(stat='identity', aes(fill=THRA1_higher), width=.5)  +
   theme_light(base_size = 12)+
  scale_fill_manual(name="Predominant Isoform", 
                    labels = c("THRA1", "THRA2"), 
                    values = c("TRUE"="darksalmon", "FALSE"="#56B4E9")) + 
  scale_y_continuous("% THRA1" ,limits=c(0, 100))+ 
  geom_hline(yintercept=50, linetype="dashed", color = "black")+
  labs(subtitle="Percentage of THRA1 expression over the total of THRA", 
       title= "THRA isoform expression pattern") + 
    theme(axis.title.y=element_blank()) +
  coord_flip()

ggsave("THRA1vsA2_mosdepth_percentages.png" , device=png, dpi = 600, path = path_plots, bg = "transparent", width = 20, height = 15, units = "cm" )



# Diverging Barcharts
ggplot(heatmap_data_percentage, aes(x=Sample, y=delta_percentage, label=delta_percentage)) + 
  geom_bar(stat='identity', aes(fill=THRA1_higher), width=.5)  +
   theme_light(base_size = 12)+
  scale_fill_manual(name="Predominant Isoform", 
                    labels = c("THRA1", "THRA2"), 
                    values = c("TRUE"="darksalmon", "FALSE"="#56B4E9")) + 
  scale_y_continuous("Delta THRA1-THRA2 in %" ,limits=c(-100, 100))+
  labs(subtitle="Delta THRA1-THRA2 expressed in percentage", 
       title= "THRA isoform expression pattern") + 
    theme(axis.title.y=element_blank()) +
  coord_flip()



```

```{r}
### Lets add a graph where the replicates are visible
heatmap_data_percentage_boxplot <- filter(dataset_final, Package == "mosdepth", sample_type == "tissue") %>%
  select(Sample, THRA1_Percentage)  %>%
  group_by(Sample) %>%
  mutate(THRA1_Percentage_mean=round(mean(THRA1_Percentage))) %>%
  mutate("THRA1_higher" = THRA1_Percentage_mean >50 ) %>%
  mutate(replicates = sum(Sample == Sample)) %>%  ## counts the repicates (the n number)
  mutate(labels = paste0(Sample, ", n=", replicates)) %>%
  ungroup()

heatmap_data_percentage_boxplot <- heatmap_data_percentage_boxplot[order(heatmap_data_percentage_boxplot$THRA1_Percentage_mean), ]  # sort
heatmap_data_percentage_boxplot <- mutate(heatmap_data_percentage_boxplot, ID = row_number())

# Diverging Barcharts
ggplot(heatmap_data_percentage_boxplot, aes(x=reorder(labels, ID), y=THRA1_Percentage, label=THRA1_Percentage)) + 
  geom_boxplot(aes(fill=THRA1_higher), width=.5)  +
  #  geom_jitter(color="black", size=0.9, alpha=0.9) +  ## optional in case of a lot of replicates to avoid overlapping of dots
   theme_light(base_size = 12)+
  scale_fill_manual(name="Predominant Isoform", 
                    labels = c("THRA1", "THRA2"), 
                    values = c("TRUE"="darksalmon", "FALSE"="#56B4E9")) + 
  scale_y_continuous("% THRA1" ,limits=c(0, 100))+ 
  geom_hline(yintercept=50, linetype="dashed", color = "black")+
  labs(subtitle="Percentage of THRA1 expression over the total of THRA", 
       title= "THRA isoform expression pattern") + 
    theme(axis.title.y=element_blank()) +
  coord_flip()


ggsave("THRA1vsA2_mosdepth_percentages_boxplot.png" , device=png, dpi = 600, path = path_plots, bg = "transparent", width = 20, height = 15, units = "cm" )

```


```{r}
### heatmap with continuos delta THRA1vsA2 using bedtools
heatmap_data_bedtools <- filter(dataset_final, Package == "bedtools")%>%
  group_by(Sample, Package) %>%
  summarize(delta_A1vsA2=round(mean(delta_A1vsA2)))

y_max_bedtools <- max(heatmap_data_bedtools$delta_A1vsA2)
y_min_bedtools <- min(heatmap_data_bedtools$delta_A1vsA2)

ggplot(heatmap_data_bedtools, aes(1, reorder(Sample,delta_A1vsA2), fill=delta_A1vsA2)) +
  geom_tile()+
  ggtitle("THRA isoform expression pattern (THRA1/THRA2) - bedtools") +
  theme_light(base_size = 12)+
  theme(axis.title.y=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_fill_gradient2('delta_reads', limits=c(y_min_bedtools, y_max_bedtools), breaks = c( y_min_bedtools, 0, y_max_bedtools),  low = "#56B4E9", high = "darksalmon", guide="colorbar")

ggsave("Heatmap_THRA1vsA2_bedtools.png" , device=png, dpi = 600, path = path_plots, bg = "transparent", width = 20, height = 15, units = "cm" )
```





```{r}
### heatmap with AUT AUT delta THRA1vsA2 using bedtools
heatmap_data_bedtools <- heatmap_data_bedtools %>%
  mutate("THRA1_higher" = delta_A1vsA2 > 0)

ggplot(heatmap_data_bedtools, aes(1, reorder(Sample,delta_A1vsA2), fill=THRA1_higher)) +
  geom_tile()+
  ggtitle("THRA isoform expression pattern (THRA1/THRA2) - bedtools") +
  theme_light(base_size = 12)+
  theme(axis.title.y=element_blank(),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_fill_manual(values = c("TRUE"= "darksalmon", "FALSE" = "#56B4E9"))

ggsave("Heatmap_THRA1vsA2_AUT_AUT_bedtools.png" , device=png, dpi = 600, path = path_plots, bg = "transparent", width = 20, height = 15, units = "cm" )
```



```{r}
###Lets compare the methods
heatmap_data_comparison <- select(dataset_final, Sample, Package, delta_A1vsA2)%>%
  group_by(Sample, Package) %>%
  summarize(delta_A1vsA2=round(mean(delta_A1vsA2)))

ggplot(heatmap_data_comparison, aes(Package, reorder(Sample,`delta_A1vsA2`) , fill=`delta_A1vsA2`)) +
  geom_tile()+
  ggtitle("Delta comparison between methods (THRA1/THRA2)") +
  theme_light(base_size = 12)+
  theme(axis.title.y=element_blank()) +
  theme(axis.title.x=element_blank()) +
  scale_fill_gradient2('delta', breaks = c(min(y_min_bedtools, y_min_mosdepth), 0,  max(y_max_bedtools,y_max_mosdepth )),  low = "#56B4E9", high = "darksalmon", guide="colorbar")


ggsave("Heatmap_methods_comparison.png" , device=png, dpi = 600, path = path_plots, bg = "transparent", width = 20, height = 15, units = "cm" )
```

